sdk/java/dagger-java-annotation-processor/src/main/java/io/dagger/annotation/processor/DaggerModuleAnnotationProcessor.java

```diff
@@ -0,0 +1,444 @@
+            if (hasDescription || parameterInfo.optional()) {
+              rm.addCode(", new $T.WithArgArguments()", io.dagger.client.Function.class);
+              if (hasDescription) {
+                rm.addCode(".withDescription($S)", parameterInfo.description());
+              }
+              if (hasDefaultValue) {
+                rm.addCode(
+                    ".withDefaultValue($T.from($S))", JSON.class, parameterInfo.defaultValue());
+              }
+            }
+            rm.addCode(")");
+          }
+          rm.addCode(")"); // end of .withFunction(
+        }
+        rm.addCode(")"); // end of .withObject(
+      }
+      rm.addCode(";\n") // end of module instantiation
+          .addStatement("return module.id()");
+
+      var im =
+          MethodSpec.methodBuilder("invoke")
+              .addModifiers(Modifier.PRIVATE)
+              .returns(JSON.class)
+              .addException(Exception.class)
+              .addParameter(JSON.class, "parentJson")
+              .addParameter(String.class, "parentName")
+              .addParameter(String.class, "fnName")
+              .addParameter(
+                  ParameterizedTypeName.get(Map.class, String.class, JSON.class), "inputArgs")
+              .beginControlFlow("try (var jsonb = $T.create())", JsonbBuilder.class);
+      var firstObj = true;
+      for (var objectInfo : moduleInfo.objects()) {
+        if (firstObj) {
+          firstObj = false;
+          im.beginControlFlow("if (parentName.equals($S))", objectInfo.name());
+        } else {
+          im.nextControlFlow("else if (parentName.equals($S))", objectInfo.name());
+        }
+        im.addStatement("$T clazz = Class.forName($S)", Class.class, objectInfo.qualifiedName())
+            .addStatement("var obj = $T.fromJSON(dag, parentJson, clazz)", JsonConverter.class)
+            .addStatement(
+                "clazz.getMethod(\"setClient\", $T.class).invoke(obj, dag)", Client.class);
+        var firstFn = true;
+        for (var fnInfo : objectInfo.functions()) {
+          var fnReturnClass = classForName(fnInfo.returnType());
+          if (firstFn) {
+            firstFn = false;
+            im.beginControlFlow("if (fnName.equals($S))", fnInfo.name());
+          } else {
+            im.nextControlFlow("else if (fnName.equals($S))", fnInfo.name());
+          }
+          var fnBlock =
+              CodeBlock.builder().add("$T fn = clazz.getMethod($S", Method.class, fnInfo.qName());
+          var invokeBlock =
+              CodeBlock.builder().add("$T res = ($T) fn.invoke(obj", fnReturnClass, fnReturnClass);
+          for (var parameterInfo : fnInfo.parameters()) {
+            Class<?> paramClazz = classForName(parameterInfo.type());
+            fnBlock.add(", $T.class", paramClazz);
+
+            invokeBlock.add(", $L", parameterInfo.name());
+
+            im.addStatement("$T $L = null", paramClazz, parameterInfo.name());
```

Comment:
This statement does not work with primitive types (boolean, int, etc). You cannot set a primitive typed var to null

Reply:
I suggest
```
            if (paramClazz.isPrimitive()) {
              if (paramClazz == int.class) {
                im.addStatement("$T $L = 0", paramClazz, parameterInfo.name());
              } else if (paramClazz == boolean.class) {
                im.addStatement("$T $L = false", paramClazz, parameterInfo.name());
              }
            } else {
              im.addStatement("$T $L = null", paramClazz, parameterInfo.name());
            }
```

Reply:
we will probably have to handle other primitive types like byte, long, char (and maybe float and double) but we can do it a future PR

Reply:
Fixed, thanks :-)